app/src/main/java/org/adaway/helper/ImportExportHostsHelper.java

```diff
@@ -0,0 +1,294 @@
+            if (results.length < 1) {
+                return null;
+            }
+            // Get URI to export lists
+            Uri result = results[0];
+            // Get context from weak reference
+            Context context = this.mWeakContext.get();
+            if (context == null) {
+                return null;
+            }
+            // Get database
+            AppDatabase database = AppDatabase.getInstance(context);
+            // Get input stream from user selected URI
+            try (InputStream inputStream = context.getContentResolver().openInputStream(result)) {
+                if (inputStream != null) {
+                    // Create reader from input stream
+                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+                        Set<String> newHosts = new HashSet<>();
+                        String line = reader.readLine();
+                        while (line != null) {
+                            if (HostsSource.isValidUrl(line)) {
+                                newHosts.add(line);
+                            }
+                            // read next line
+                            line = reader.readLine();
+                        }
+                        this.importHosts(database, newHosts);
+                    }
+                }
+            } catch (FileNotFoundException exception) {
+                Log.e(Constants.TAG, "File not found!", exception);
+            } catch (IOException exception) {
+                Log.e(Constants.TAG, "IO Exception", exception);
+            }
+            // Return nothing
+            return null;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+            // Get context from weak reference
+            Context context = this.mWeakContext.get();
+            if (context == null) {
+                return;
+            }
+            // Check and show progress dialog
+            this.mProgressDialog = new ProgressDialog(context);
+            this.mProgressDialog.setMessage(context.getString(R.string.import_dialog));
+            this.mProgressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
+            this.mProgressDialog.setCancelable(false);
+            this.mProgressDialog.show();
+        }
+
+        @Override
+        protected void onPostExecute(Void unused) {
+            super.onPostExecute(unused);
+            // Check progress dialog
+            if (this.mProgressDialog != null) {
+                this.mProgressDialog.dismiss();
+            }
+        }
+
+        private void importHosts(AppDatabase database, Set<String> hosts) {
+            for (String url: hosts) {
+                HostsSource source = new HostsSource();
+                source.setUrl(url);
+                source.setEnabled(true);
+                AppExecutors.getInstance().diskIO().execute(() -> database.hostsSourceDao().insert(source));
```

Comment:
What happens if the source already exists? I mean, does it crash?

Reply:
No, it handles it.
I tested via creating an export, adding a new host onto the export, and then importing the new file and it was fine. No crash, and no duplicates